arangod/Aql/OptimizerRulesReplaceFunctions.cpp

```diff
@@ -0,0 +1,467 @@
+
+  /// index
+  //  we create this first as creation of this node is more
+  //  likely to fail than the creation of other nodes
+
+  //  index - part 1 - figure out index to use
+  std::shared_ptr<arangodb::Index> index = nullptr;
+	std::vector<basics::AttributeName> field;
+	TRI_ParseAttributeString(params.attribute, field, false);
+	auto indexes = trx->indexesForCollection(params.collection);
+	for(auto& idx : indexes){
+		if(idx->type() == arangodb::Index::IndexType::TRI_IDX_TYPE_FULLTEXT_INDEX) {
+			if(basics::AttributeName::isIdentical(idx->fields()[0], field, false /*ignore expansion in last?!*/)) {
+				index = idx;
+				break;
+			}
+		}
+	}
+
+	if(!index){ // not found or error
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_QUERY_FULLTEXT_INDEX_MISSING);
+	}
+
+  // index part 2 - get remaining vars required for index creation
+  auto* aqlCollection = aql::addCollectionToQuery(query, params.collection, false);
+  if(!aqlCollection) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_QUERY_FUNCTION_ARGUMENT_TYPE_MISMATCH,"collection used in FULLTEXT not found");
+  }
+	auto condition = std::make_unique<Condition>(ast);
+	condition->andCombine(funAstNode);
+  condition->normalize(plan);
+	// create a fresh out variable
+  Variable* indexOutVariable = ast->variables()->createTemporaryVariable();
+
+  ExecutionNode* eIndex = plan->registerNode(
+    new IndexNode(
+      plan,
+      plan->nextId(),
+      aqlCollection,
+      indexOutVariable,
+      std::vector<transaction::Methods::IndexHandle> {
+				transaction::Methods::IndexHandle{index}
+      },
+      std::move(condition),
+      IndexIteratorOptions()
+    )
+  );
+
+  //// wrap plan part into subquery
+  return createSubqueryWithLimit(plan, calcNode, eIndex, eIndex, indexOutVariable, params.limit);
+}
+
+} // namespace
+
+
+
+//! @brief replace legacy JS Functions with pure AQL
+void arangodb::aql::replaceNearWithinFulltext(Optimizer* opt
+                                             ,std::unique_ptr<ExecutionPlan> plan
+                                             ,OptimizerRule const* rule){
+
+  bool modified = false;
+
+  SmallVector<ExecutionNode*>::allocator_type::arena_type a;
+  SmallVector<ExecutionNode*> nodes{a};
+  plan->findNodesOfType(nodes, ExecutionNode::CALCULATION, true);
+
+  for(auto const& node : nodes){
+    auto visitor = [&modified, &node, &plan](AstNode* astnode){
+      auto* fun = getFunction(astnode); // if fun != nullptr -> astnode->type NODE_TYPE_FCALL
+      AstNode* replacement = nullptr;
+      if(fun){
+        if (fun->name == "NEAR"){
+          replacement = replaceNearOrWithin(astnode, node, plan.get(), true /*isNear*/);
+        } else if (fun->name == "WITHIN"){
+          replacement = replaceNearOrWithin(astnode, node, plan.get(), false /*isNear*/);
+        } else if (fun->name == "FULLTEXT"){
+          replacement = replaceFullText(astnode, node,plan.get());
+        }
+      }
+      if (replacement) {
```

Comment:
does this now make sure all occurrences of NEAR/WITHIN/FULLTEXT are actually replaced?

Reply:
answer: it does, but I had to scroll around between 3 different functions to figure this out.